pynamodb/attributes.py

```diff
@@ -471,16 +468,38 @@ def serialize(self, values):
 
     def deserialize(self, values):
         """
-        Decode numbers from list of AttributeValue types.
+        Decode as a dict.
         """
         deserialized_dict = dict()
         for k in values:
             v = values[k]
-            attr_class = _get_class_for_deserialize(v)
             attr_value = _get_value_for_deserialize(v)
-            deserialized_dict[k] = attr_class.deserialize(attr_value)
+            key = self._dynamo_to_python_attr(k)
+            attr_class = self._get_deserialize_class(key, v)
+
+            # need to see if this is a typed map and use the concrete type as `attr_class` here
+            deserialized_dict[key] = attr_class.deserialize(attr_value)
+
+        # If this is a subclass of a MapAttribute (i.e typed), instantiate an instance
+        if not self.is_raw():
+            return type(self)(**deserialized_dict)
         return deserialized_dict
 
+    @classmethod
+    def is_raw(cls):
+        return cls == MapAttribute
```

Comment:
I don't love this, but I don't know of another way of doing this without catching subclasses, since `issubclass(MapAttribute, MapAttribute)` also returns True.